---
title: "Gene Expression Analysis"
author: "Viet Tran"
format: pdf
editor: visual
---

# Research Questions

```{r}
library(dplyr)
library(tidyr)
library(DESeq2)
```

```{r}
downloads_path  = '/Users/luke/Desktop/UCD/Class/Bio Principle'
file_path = paste(downloads_path,"brca_tcga_pan_can_atlas_2018.tar.gz", sep = "/" )

untar(file_path)

folder_path = paste(getwd(),"brca_tcga_pan_can_atlas_2018", sep = "/" )
```

# Literature Review

```{r}
rna_seq_path = paste(folder_path,"data_mrna_seq_v2_rsem.txt", sep = "/")
patient_data_path = paste(folder_path,"data_clinical_patient.txt", sep = "/")
copy_number_aberrations_path = paste(folder_path,"data_cna.txt", sep = "/")

rna_seq = read.delim(rna_seq_path)
data_patient = read.delim(patient_data_path)
cna = read.delim(copy_number_aberrations_path)
```

```{r}
str(rna_seq)
str(cna)
str(data_patient)
```

```{r}
# Remove the first 4 rows of data_patient
data_patient = data_patient[5:1088,]

row.names(data_patient) <- 1:nrow(data_patient)
head(data_patient)
```

```{r}
col_age = which(colnames(data_patient)=='Diagnosis.Age')
# set plots to come side by side


# par(mfrow = c(1, 3))
# Get histogram of age
hist(as.numeric(data_patient[,col_age]), main = "Histogram of Age at Diagnosis", breaks = 50, xlab = "Age at Diagnosis")

# Get Histogram of Stage
col_stage = which(colnames(data_patient) == "Neoplasm.Disease.Stage.American.Joint.Committee.on.Cancer.Code")

# For simplicity remove the A,B, C, subcategories. 

stage = data_patient[,col_stage]

# remove A

stage = gsub("I[^VI]", "I", stage)
stage = gsub("V[^\\s]", "V", stage)


barplot(table(as.factor(stage)),  main = "Histogram of Diagnosed Stage", ylab="Frequency", las = 2, cex.names = 1 )
# Get Histogram of (Uncensored) Survival Times 
col_OS = which(colnames(data_patient) == "Overall.Survival.Status")

uncensored = which(data_patient[,col_OS] == "1:DECEASED")

col_OS_T = which(colnames(data_patient) == "Overall.Survival..Months.")
hist(as.numeric(data_patient[uncensored,col_OS_T]), breaks = 50, main = "Histogram of Overall survival Months", xlab = "Survival (in months)")

mean(as.numeric(data_patient[uncensored,col_OS_T]))
```

```{r}
assay = round(as.matrix(rna_seq[, -c(1,2)]))
rownames(assay) = rna_seq[, 1]

metadata = matrix(0, dim(assay)[2], 3)

pat_ids = data_patient[,1]

stage_factor = factor(stage, ordered = T, levels = c("STAGE I", "STAGE II","STAGE III","STAGE IV", "STAGE X", NA)) # Need to remove stage X and NA

cna_long = cna |>
  pivot_longer(cols = -c(Hugo_Symbol, Entrez_Gene_Id),
               names_to = "Patient_ID",
               values_to = "CNA")
amplified_patient = cna_long |>
  filter(Hugo_Symbol == "ERBB2", CNA > 0)

amplified_patient_id = amplified_patient$Patient_ID

clean_id <- function(x) {
  x <- substr(x, 1, 12)
  x <- gsub("\\.", "-", x)
  x
}

amplified_patient_id = clean_id(amplified_patient_id)
colnames(assay) = clean_id(colnames(assay))

for (i in 1:dim(assay)[2]) {
  pat_barcode = colnames(assay)[i]
  
  idx = which(pat_barcode == pat_ids)
  
  metadata[i, 1] = 1*(as.numeric(data_patient[idx,col_age])<55)
  metadata[i,2] = stage_factor[idx]
  metadata[i,3] = ifelse(pat_barcode %in% amplified_patient_id, 1, 0)
}

metadata[is.na(metadata)] = 0
metadata[metadata[,2] == 5, 2] = 0
colnames(metadata) = c("Early", "Stage", "ERBB2")
```

```{r}
assay[is.na(assay)] = 0
assay[assay < 0] = 0

smallestGroupSize = 3
keep = rowSums(assay >= 10) >= smallestGroupSize

assay = assay[keep,]

dds = DESeqDataSetFromMatrix(countData = assay,
                              colData = metadata,
                              design = ~ ERBB2)
```

```{r}
dds = DESeq(dds)
resultsNames(dds)

res = results(dds, alpha = 1e-5)

significant_res = res[res$padj < 1e-5 & abs(res$log2FoldChange) > 0 , ]

significant_res[order(abs(significant_res$log2FoldChange), decreasing = TRUE) & order(significant_res$padj), ]

significant_res[order(significant_res$padj)[1:10], ]

hist(significant_res$log2FoldChange)

plotMA(res)
```

```{r}
vsd = vst(dds)

par(mfrow = c(1, 2))
plotPCA(vsd, intgroup=c("ERBB2"))
plotPCA(vsd, intgroup=c("Early"))
```

```{r}
library(clusterProfiler)
library(enrichplot)
library(org.Hs.eg.db)
```

```{r}
res_sig = res[res$padj<0.05,]

# separate into over and under expressed using log2foldchange
DE_over = rownames(res_sig[res_sig$log2FoldChange>0,])
DE_under = rownames(res_sig[res_sig$log2FoldChange<0,])

go_results_over = enrichGO(
  gene          = DE_over,
  OrgDb         = org.Hs.eg.db,
  keyType       = "SYMBOL",  
  ont           = "BP", 
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.05,
  qvalueCutoff  = 0.05
)

# print and plot results
print(head(go_results_over))



go_results_under = enrichGO(
  gene          = DE_under,
  OrgDb         = org.Hs.eg.db,
  keyType       = "SYMBOL",  
  ont           = "BP", 
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.05,
  qvalueCutoff  = 0.05
)

# print and plot results
print(head(go_results_under))

dotplot(go_results_under, showCategory=10) + ggtitle("Gene Ontology Enrichment Under Expressed")
```

```{r}
library(ReactomePA)
library(pathview)

gene_entrez_over <- bitr(
  DE_over,
  fromType = "SYMBOL",
  toType   = "ENTREZID",
  OrgDb    = org.Hs.eg.db
)


gene_entrez_over <- bitr(
  DE_over,
  fromType = "SYMBOL",
  toType   = "ENTREZID",
  OrgDb    = org.Hs.eg.db
)

gene_entrez_under <- bitr(
  DE_under,
  fromType = "SYMBOL",
  toType   = "ENTREZID",
  OrgDb    = org.Hs.eg.db
)

kegg_results_over =  enrichKEGG(
  gene          = gene_entrez_over[,2],
  organism      = "human",   
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.05,
  qvalueCutoff  = 0.05
)

kegg_results_under =  enrichKEGG(
  gene          = gene_entrez_under[,2],
  organism      = "human",   
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.05,
  qvalueCutoff  = 0.05
)

print(head(kegg_results_over))

dotplot(kegg_results_over, showCategory=10) + ggtitle("Kegg Pathway Enrichment Over Expressed")

print(head(kegg_results_under))

dotplot(kegg_results_under, showCategory=10) + ggtitle("Kegg Pathway Enrichment Under Expressed")
```

```{r}
reactome_results_over =  enrichPathway(
  gene          = gene_entrez_over[,2],
  organism      = "human",   
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.05,
  qvalueCutoff  = 0.05,
)

reactome_results_under =  enrichPathway(
  gene          = gene_entrez_under[,2],
  organism      = "human",   
  pAdjustMethod = "BH",
  pvalueCutoff  = 0.05,
  qvalueCutoff  = 0.05,
)


print(head(reactome_results_over))

dotplot(reactome_results_over, showCategory=10) + ggtitle("Reactome Pathway Enrichment Over Expressed")

print(head(reactome_results_under))

dotplot(reactome_results_under, showCategory=10) + ggtitle("Reactome Pathway Enrichment Under Expressed")
```

```{r}
go_results_under_pw = pairwise_termsim(go_results_under)
treeplot(go_results_under_pw)+ ggtitle("GO Enrichment Under Expressed")

kegg_results_under_pw = pairwise_termsim(kegg_results_under)
treeplot(kegg_results_under_pw)+ ggtitle("KEGG Enrichment Under Expressed")
```

```{r}
# Subset the dataset on differentially expressed gene for this assignment. 
top_DE = order(res$padj)

vsd_DE = assay(vsd)[top_DE[1:20],]
# to simplify the visualization get top most different genes. 


# install packages for nicer heatmap than R's base one. 


library(pheatmap)


annotation_colors = list(ERBB2 = c(T = "#1f78b4", F = "#33a02c"))



annotation_col = data.frame(ERBB2 = as.matrix(metadata[,3]))
rownames(annotation_col) = colnames(vsd)


pheatmap(
  vsd_DE,
  cluster_rows = TRUE,      
  cluster_cols = TRUE,  
  scale = 'row',
  show_colnames = FALSE,
  show_rownames = TRUE,
  annotation_col = annotation_col)
```

\
